diff -ur opengl-orig/glut.c opengl/glut.c
--- opengl-orig/glut.c	2000-01-06 09:37:43.000000000 -0700
+++ opengl/glut.c	2003-09-07 09:14:48.000000000 -0600
@@ -15,7 +15,7 @@
 static VALUE _funcname = Qnil; \
 static void glut_ ## _funcname ## Callback(); \
 static VALUE \
-glut_ ## _funcname ## (obj,arg1) \
+glut_ ## _funcname (obj,arg1) \
 VALUE obj,arg1; \
 { \
     int win; \
@@ -25,7 +25,7 @@
     if (win == 0) \
 	rb_raise(rb_eRuntimeError, "GLUT.%s needs current window", #_funcname); \
     rb_ary_store(_funcname, win, arg1); \
-    glut ## _funcname ## (glut_ ## _funcname ## Callback); \
+    glut ## _funcname (glut_ ## _funcname ## Callback); \
     return Qnil; \
 }
 
Only in opengl: glut.c~
