diff -ur opengl.orig/extconf.rb opengl/extconf.rb
--- opengl.orig/extconf.rb	2001-12-01 06:24:58.000000000 +0100
+++ opengl/extconf.rb	2005-03-27 12:00:49.538754864 +0200
@@ -39,12 +39,13 @@
   gl_libname = "GL"
   glu_libname = "GLU"
   glut_libname = "glut"
+  $LIBPATH = "@LIBPATH@"
 end
 
 $objs = ["glu.o", "ogl.o", "rbogl.o"]
 #have_library("pthread", "pthread_create")
-(have_library(gl_libname, nil) || have_library("Mesa"+gl_libname, nil)) &&
-  (have_library(glu_libname, nil) ||  have_library("Mesa"+glu_libname, nil)) &&
+(have_library(gl_libname) || have_library("Mesa"+gl_libname)) &&
+  (have_library(glu_libname) ||  have_library("Mesa"+glu_libname)) &&
   create_makefile("opengl")
 if (FileTest.exist? "Makefile")
     open("Makefile") {|f|
@@ -64,7 +65,7 @@
 have_library("Xi", "XAllowDeviceEvents") &&
   have_library("Xext", "XMITMiscGetBugMode") &&
   have_library("Xmu", "XmuAddCloseDisplayHook")
-have_library(glut_libname, nil) &&
+have_library(glut_libname) &&
   create_makefile("glut")
 if (FileTest.exist? "Makefile")
     open("Makefile") {|f|
